src/RewriteInstance.cpp

```diff
@@ -2955,7 +2986,7 @@ void RewriteInstance::emitSections() {
 void RewriteInstance::emitFunctions(MCStreamer *Streamer) {
   auto emit = [&](const std::vector<BinaryFunction *> &Functions) {
     for (auto *Function : Functions) {
-      if (!BC->HasRelocations &&
+      if ((!BC->HasRelocations || Function->getSection().isProtected()) &&
```

Comment:
If there are references from functions in protected sections to functions in regular sections, how are these references going to get updated?

Reply:
I'm working on a fix for that but I have a question around it.
On this fix, I let BOLT processing the relocations on protected sections but I prevent moving the section itself, and all the functions inside of it.

But I still have in issue. BOLT is generating an other section (```.org.boltcustom```) to carry my ```__start_custom``` and my ```__stop_custom``` symbol.
Beside not  to be the reaction I want, this section is a perfect no-sense, as this section begin at ```__start_custom``` and have a size of 0. On the other hand, even if it had the "good" size, it would have been a non-sense either, as that means ```.org.boltcustom``` and ```custom``` would have been overlaid.

I tried a lot of things but I still have this section, either as ```.org.boltcustom``` or as a duplicate of ```custom```. My question is, do you know or can you give me an hint of how prevent this section creation?

Reply:
One idea is to handle immovable code sections as data. This was, the will be no need to even build `BinaryFunction`s for the contents. Of course the problem of updating the references still remains, and we have to rely on relocation info. The major exception from regular data sections is that there will be PC-relative relocations that need a special type of handling.

Reply:
Hi @maksfb,

I am helping a bit Pierre on trying to make Bolt work on a binary linked against TcMalloc. My knowledge of Bolt is far below yours and the one of Pierre ;)

We don't understand fully what you mean. From what I get, you suggest us to handle the custom section as only raw data, but not functions, ok that point is fine. However as this section shall be translated by some offset in the final binary memory layout, we have to update:
 - any code outside of custom that calls any symbol defined in custom
 - any code inside custom that calls any symbol outside custom.

So we still need:
 - to list all symbols of custom and update the "normal" code referencing them
 - to update the code of custom itself (so parse it as functions and update the corresponding assembly).

Am I right ? So this doesn't really go in the direction of considering it fully as data.

From what Pierre told me, today everything works (ie "custom" memory layout is correctly kept, still functions inside it are modified accordinly when they reference foreign symbols), just the only problem is about this "__stop_" marker which is being wrongly associated to the following section/symbol, and thus relocated elsewhere. Do you have a simple advice for us on how exactly we should tackle this corner case in Bolt. Or any other clue we should try to dig in ? We aren't ELD experts as we would like to, and I know by experience explaining all these things takes a lot of time, but we will take all the clues you might give us to help us going in the right direction ;)

Cheers,
Romain

Reply:
Hi Romain and Pierre,

It's hard for me to tell exactly what is going on without looking at your other patch.

`.bolt.orgcustom` should not get created unless we are relocating `custom` section. With the current patch I cannot reproduce this problem, i.e. `custom` section stays in place.

You are correct listing items that need to be implemented. If you have almost everything working without it, then it might be a simpler route to take.

The `__stop` marker needs a special handling in `updateSymbolTable()`. If you can recognize it and skip its update then it should be enough.

Hope this helps. Again, it would be easier for me to assist you if I could reproduce the issue with the other patch you are talking about.